evennia/contrib/game_systems/containers/containers.py

```diff
@@ -0,0 +1,301 @@
+"""
+Containers
+
+Contribution by InspectorCaracal (2023)
+
+Adds the ability to put objects into other container objects by providing a container typeclass and extending certain base commands.
+
+To install, import and add the `ContainerCmdSet` to `CharacterCmdSet` in your `default_cmdsets.py` file:
+
+    from evennia.contrib.game_systems.containers import ContainerCmdSet
+
+    class CharacterCmdSet(default_cmds.CharacterCmdSet):
+        # ...
+        
+        def at_cmdset_creation(self):
+            # ...
+            self.add(ContainerCmdSet)
+
+The ContainerCmdSet includes:
+
+ - a modified `look` command to look at or inside objects
+ - a modified `get` command to get objects from your location or inside objects
+ - a new `put` command to put objects from your inventory into other objects
+
+Create objects with the `ContribContainer` typeclass to easily create containers,
+or implement the same locks/hooks in your own typeclasses.
+
+`ContribContainer` implements the following new methods:
+
+    at_pre_get_from(getter, target, **kwargs) - called with the pre-get hooks
+    at_pre_put_in(putter, target, **kwargs)   - called with the pre-put hooks
+"""
+from django.conf import settings
+
+from evennia import AttributeProperty, CmdSet, DefaultObject
+from evennia.commands.default.general import CmdLook, CmdGet, CmdDrop
+from evennia.utils import class_from_module
+
+# establish the right inheritance for container objects
+_BASE_OBJECT_TYPECLASS = class_from_module(settings.BASE_OBJECT_TYPECLASS, DefaultObject)
+
+
+class ContribContainer(_BASE_OBJECT_TYPECLASS):
+    """
+    A type of Object which can be used as a container.
+
+    It implements a very basic "size" limitation that is just a flat number of objects.
+    """
+
+    # This defines how many objects the container can hold.
+    capacity = AttributeProperty(default=20)
+
+    def at_object_creation(self):
+        """
+        Extends the base object `at_object_creation` method by setting the "get_from" lock to "true",
+        allowing other objects to be put inside and removed from this object.
+
+        By default, a lock type not being explicitly set will fail access checks, so objects without
+        the new "get_from" access lock will fail the access checks and continue behaving as
+        non-container objects.
+        """
+        super().at_object_creation()
+        self.locks.add("get_from:true()")
+
+    def at_pre_get_from(self, getter, target, **kwargs):
+        """
+        This will be called when something attempts to get another object FROM this object,
+        rather than when getting this object itself.
+
+        Args:
+            getter (Object): The actor attempting to take something from this object.
+            target (Object): The thing this object contains that is being removed.
+
+        Returns:
+            boolean: Whether the object `target` should be gotten or not.
```

Comment:
It would be good to explain what this means; will the get operation be quietly aborted or will an error message be reported automatically - or is that something that should be handled in this method too?

Reply:
I've added `Notes` to this and the other hook matching the extant methods explaining that the action is aborted, but I'm unsure what the correct answer regarding the error messages should be, since the core code is equally unclear on the matter.